wasm/src/signature.rs

```diff
@@ -0,0 +1,149 @@
+use contracts_common::custom_serde::BytesSerializable;
+use ethers::utils::keccak256;
+use js_sys::{Function, Promise};
+use num_bigint::BigUint;
+use wasm_bindgen::prelude::*;
+use wasm_bindgen_futures::JsFuture;
+
+use crate::{
+    circuit_types::transfers::{
+        to_contract_external_transfer, ExternalTransfer, ExternalTransferDirection,
+    },
+    common::{derivation::derive_sk_root_scalars, types::Wallet},
+    helpers::{bytes_from_hex_string, bytes_to_hex_string},
+    sign_commitment,
+};
+
+/// Signs a wallet commitment based on key type.
+///
+/// Internal type uses seed to derive signing key.
+/// External type uses provided sign_message function.
+pub async fn sign_wallet_commitment(
+    wallet: &Wallet,
+    seed: &str,
+    key_type: &str,
+    sign_message: Option<&Function>,
+) -> Result<Vec<u8>, String> {
+    let comm = wallet.get_wallet_share_commitment();
+
+    match key_type {
+        "internal" => {
+            let old_sk_root = derive_sk_root_scalars(seed, &wallet.key_chain.public_keys.nonce);
+            Ok(sign_commitment(&old_sk_root, comm)
+                .map_err(|e| e.to_string())?
+                .to_vec())
+        }
+        "external" => {
+            let sign_message = sign_message.ok_or_else(|| {
+                String::from("sign_message function is required for external key type")
+            })?;
+            generate_signature(wallet, sign_message).await
+        }
+        _ => Err(String::from("Invalid key type")),
+    }
+}
+
+/// Generates a signature by calling the provided sign_message function
+pub async fn generate_signature(
+    wallet: &Wallet,
+    sign_message: &Function,
+) -> Result<Vec<u8>, String> {
+    let comm_bytes = wallet
+        .get_wallet_share_commitment()
+        .inner()
+        .serialize_to_bytes();
+
+    let digest = keccak256(comm_bytes);
+    let digest_hex = bytes_to_hex_string(&digest);
```

Comment:
don't think we'll want to prehash the commitment, the signature APIs of MPC wallet providers probably don't expect a prehashed message